### PR TITLE
fix(coverage): respect shell redirection in extractor (#40)

### LINKED ISSUE #40: dippin coverage flags file-redirected echo/printf as 'uncovered tool outputs'
## Summary

The tool-output extraction in `coverage/coverage.go` matches every `printf '...'` and `echo '...'` in a shell command body, regardless of whether that statement redirects to a file. This produces false-positive "uncovered output" entries for log writes that never reach stdout, causing healthy pipelines to drop from A/100 to A/90 in `dippin doctor`.

## Concrete repro

The following tool node writes only `validation_pass` or `validation_fail` to stdout. Every other `echo` redirects to a log file:

```dippin
tool RunTests
  timeout: 2m
  command: |
    set -eu
    echo '=== capturing test baseline ===' > /tmp/log
    echo 'build_system=python' >> /tmp/log
    uv run pytest >> /tmp/log 2>&1
    if [ "$?" -eq 0 ]; then printf 'validation_pass'; else printf 'validation_fail'; fi
```

Edge wiring:
```dippin
edges
  RunTests -> Pass when ctx.tool_stdout = validation_pass
  RunTests -> Fail when ctx.tool_stdout = validation_fail
```

The conditions cover every value the tool can possibly emit to `stdout`. But `dippin coverage RunTests.dip` reports:

```
✗ RunTests  partial
    missing: validation_pass
    missing: validation_fail
    missing: === capturing test baseline ===
    missing: build_system=python
```

(The `validation_pass` / `validation_fail` rows might also appear if `\n` was added to the printf and edge conditions don't include the newline — separate concern.)

The three echoes that redirect to `/tmp/log` should NOT be counted as possible stdout values. They never reach stdout.

## Root cause

`coverage/coverage.go:48-53`:

```go
// outputPatterns matches printf/echo string literals in shell commands.
var outputPatterns = []*regexp.Regexp{
    regexp.MustCompile(`printf\s+'([^']+)'`),
    regexp.MustCompile(`printf\s+"([^"]+)"`),
    regexp.MustCompile(`printf\s+'%s'\s+'([^']+)'`),
    regexp.MustCompile(`echo\s+'([^']+)'`),
}
```

`extractToolOutputs(command string)` (around line 145) runs these regexes over the entire command body. There's no shell parsing of the surrounding context, so a statement like `echo 'foo' >> bar.txt` is treated identically to `echo 'foo'` from a stdout perspective.

## Suggested fix

Several escalating options:

### Cheapest: post-filter to drop matches whose statement contains a redirect

After regex match, scan the same line (or up to the next newline / `;` / `&&` / `||`) for `>`, `>>`, `|`. If found, treat that match as redirected and exclude it. Quick heuristic; handles the common cases (`echo X > file`, `echo X >> file`, `echo X | tee file`).

```go
func isRedirected(command, match string) bool {
    idx := strings.Index(command, match)
    if idx < 0 {
        return false
    }
    // Find the end of this statement
    lineEnd := strings.IndexAny(command[idx:], "\n;")
    if lineEnd < 0 {
        lineEnd = len(command) - idx
    }
    rest := command[idx : idx+lineEnd]
    return strings.ContainsAny(rest, ">|")
}
```

False positives in this heuristic: `printf '%s > %s'` where `>` is part of the format string. But these are rare and benign — they would just under-extract, not over-extract.

### More principled: use a shell parser

Pull in `mvdan.cc/sh/v3/syntax` or similar to parse the command body and walk the AST, looking specifically for echo/printf calls whose redirections are empty. Bigger surface but correct.

### Workaround for authors today

Declare expected outputs explicitly:

```dippin
tool RunTests
  outputs: validation_pass,validation_fail
  command: |
    ...
```

`coverage/coverage.go:99-102` already preferences declared `cfg.Outputs` over regex extraction:

```go
outputs := cfg.Outputs
if len(outputs) == 0 {
    outputs = extractToolOutputs(cfg.Command)
    cov.ExtractedOutputs = outputs
}
```

So `outputs:` is the documented escape hatch for files affected today. But many existing `.dip` files don't declare it and shouldn't have to.

## Tests to write

In `dippin-lang/coverage/coverage_test.go`:

1. Tool with `echo 'foo' > log.txt; printf 'green'` — assert `extractToolOutputs` returns `["green"]`, NOT `["foo", "green"]`.
2. Tool with `echo 'foo' >> log.txt; echo 'bar' | tee /tmp/x; printf 'green'` — assert returns `["green"]`.
3. Tool with `printf '%s\n' 'inline'; printf 'tail'` — assert returns `["inline", "tail"]` (both reach stdout).
4. Tool with `echo 'foo'` (no redirect) — assert returns `["foo"]` (regression test for normal case).

## Discovered while

Running a quality pass on `2389-research/pipelines` (PR #13). `build-and-ship/refactor-express.dip` post-TRK101-fix is otherwise clean but drops to A/90 because of file-redirected echoes in `SnapshotTests` that the extractor treats as missing edge targets.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by clintecker**
## Summary

- Replace the regex-based `extractToolOutputs` in `coverage/coverage.go` with an AST walker built on `mvdan.cc/sh/v3/syntax` (already a project dependency, already used by `validator/lint_tool_helpers.go:extractBinary`).
- Statements that redirect to files (`>`, `>>`, `&>`, `>&`), feed into pipes, or nest `echo`/`printf` inside command substitution are now skipped — they don't reach the engine's stdout.
- Add `examples/coverage_redirection.dip` exercising the bug pattern from issue #40 and verifying the `RunTests` node reports `covered`.
- Tag v0.30.0 after merge.

## Test plan

- [x] `just check` green locally (pre-commit hook ran build, vet, golangci-lint, gofmt, test-race, complexity, validate-examples, release checks)
- [ ] CI green
- [x] `examples/coverage_redirection.dip` lints clean and reports \`covered\`
- [ ] Spot-check on \`2389-research/pipelines\` PR #13 — confirm \`build-and-ship/refactor-express.dip\` flips from A/90 to A/100

Closes #40.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/dippin-lang/pr/50"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1781982459&installation_id=131176874&pr_number=50&repository=2389-research%2Fdippin-lang&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Fdippin-lang%2Fpull%2F50&signature=3c734e282d895121aa43f44c32c28175e491ccdf96107d12d82f152122a4de6a"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Coverage extractor now correctly respects shell redirections, file writes, command substitutions, and pipelines so reported tool output reflects actual stdout seen by the engine
  * Pipelines using the "log to file, printf marker on stdout" pattern are reclassified as covered

* **New Features**
  * Added an example workflow demonstrating coverage redirection behavior

* **Documentation**
  * Changelog updated for v0.30.0 describing the coverage fixes

[Review Change Stack](https://app.coderabbit.ai/change-stack/2389-research/dippin-lang/pull/50?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: examples/coverage_redirection.dip [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Regression for issue #40: echoes redirected to a file must not count
+# as stdout coverage; only the final printf should reach the engine.
+workflow CoverageRedirection
+  goal: "Demonstrate that file-redirected echoes no longer trip coverage"
+  start: RunTests
+  exit: Done
+
+  tool RunTests
+    timeout: 2m
```
[the remark this answers is not in the document]
Comment (clintecker) [reply examples/coverage_redirection.dip:9]: Verified false positive: 0 of the 27 files in `examples/` declare a top-level `defaults` block — same false-claim pattern as v0.29.0. The `defaults:` block is optional per the language grammar; `just lint-examples` produces zero DIP fires against this file. Passing on this comment.
